File: project 1/qwrap.m
% qwrap.m - vectorized version of qwrap.c
%
% Usage: q_out = qwrap(D, q) = mod-(D+1) version of q
%
% D = maximum order, buffer length = D+1
% q = circular index into buffer, must be in the range -1<=q<=2*D
%
% q_out = wrapped version of q
%
% easier by anonymous function: Q = @(D,q) q + (D+1)*((q<0) - (q>D));

% S. J. Orfanidis, 332:447 DSPD, Fall 2017

function q_out = qwrap(D, q)

if nargin==0, help qwrap; return; end

q_out = q + (D+1)*((q<0) - (q>D));

File: project 1/qwrap.c
// qwrap.c - index wrapping mod M+1 
//
// Usage: q_new = qwrap(D,q)
//
//  332:348 DSP Lab - Spring 2011 - S. J. Orfanidis
// ---------------------------------------------------------------------------- 

int qwrap(int D, int q)
{
       if (q > D)                  /* assumes q is in the bounds 0 <= q <= D */
              q -= D + 1;          /* when q=D+1, it wraps around to q=0 */

       if (q < 0)  
              q += D + 1;          /* when q=-1, it wraps around to q=D */

       return q;
}

